### ISSUE TITLE
Track the 0.2 release progress

### ISSUE THREAD

**Issue by erwin1** [pyrsia/pyrsia#1037]
This isn't a usual issue, but I needed a place to track the 0.2 progress.
This following is a list of steps that we will use to validate the 0.2 release:

- [ ] 1. setup and run an authorized node: nodeA. one way or another we need to make sure we bootstrap the blockchain/transparency log, with at least a log that marks nodeA as authorized.

--> TO BE DISCUSSED + CREATE NEW ISSUE

- [ ] 2. setup and run a regular node: nodeB
when it starts, it needs to be able to pull all existing blocks and make sure they get stored in the TransparencyLog DB 

--> #833 - Xuefei to pick up next

- [x] 3. trigger a build for a docker image:tag on nodeA

- [x] 4. nodeA uses its configured build pipeline service to start a build and wait for the result

- [ ] 5. on nodeA: when the build is ready, nodeA creates a TPL and call add_block on the Blockchain component

--> issue #1006 - Joeri will create a PR

- [x] 6. on nodeA: the blockchain needs to reach consensus with the other authorized nodes

--> I listed this step for completeness. We skip this in 0.2 - we use a 'light' consensus mechanism for now

- [ ] 7. on nodeA: the blockchain component makes sure the committed block is distributed across the Pyrsia network 

--> Will be addressed by #978 which Xuefei is finalizing

- [ ] 8. on every other node in the network including nodeA: the blockchain component listens for new blocks and notifies listeners like the TransparencyLog component, which will in turn store the log in its DB.

--> partly addressed by #1007 - needs to be investigated if that will be enough.

- [x] 9. on nodeB: do a docker pull of alpine:3.16 

- [x] 10. on nodeB: the artifact service finds a TPL for alpine:3.16 and requests the needed binaries from the p2p network (hence nodeA in this example)

- [x] 11. on nodeB: the docker pull completes successfully

- [x] 12. on any node: pyrsia inspect log docker --image image:tag outputs the transparency logs

- [ ] 13. Take steps 9 to 12 on a node behind a NAT

**Comment by betarelease** [comment]
Great this can be then converted into documentation for 0.2

**Comment by betarelease** [comment]
For 
 1. setup and run an authorized node: nodeA. one way or another we need to make sure we bootstrap the blockchain/transparency log, with at least a log that marks nodeA as authorized.
 
 Here are the steps I am thinking of 
1. Run the k8s cluster - with Pyrsia nodes
2. Inject the authority key 
3. Ensure the key is in the secure lockbox - may need to implement deploying the Vault
4. trigger a [9] action above and ensure that [6] consensus is reached.

**Comment by erwin1** [comment]
@betarelease I intended step 1-3 to be implicit in the 'setup', but it's also fine to list them explicitly. What I meant with 'bootstrapping' is making sure we start off with at least some 'AddAuthorizedNode' transactions in the transparency log (and thus blockchain) for the authorized nodes.

In theory we could skip this in 0.2 but that would mean any node could act as an authorized node and perform builds.

**Comment by betarelease** [comment]
I think adding the AddAuthorizedNode is good idea. 

That makes me think - should we add a check somewhere in the blockchain code to ensure that only an Authorizednode is committing? and force that check everytime.

**Comment by fishseabowl** [comment]
> I think adding the AddAuthorizedNode is good idea.
> 
> That makes me think - should we add a check somewhere in the blockchain code to ensure that only an Authorizednode is committing? and force that check everytime.

Yes, according to the design of the blockchain, only AuthorizedNode can reach consensus, and only consensus members can commit blocks.

**Comment by erwin1** [comment]
Acceptance criteria for 0.2 release

- [ ] a central node is running, accepting bootstrap connections via boot.pyrsia.link
- [ ] that central node initialized the blockchain
- [ ] and it added itself as the first authorized node
- [ ] a regular node bootstraps using boot.pyrsia.link by default, also behind a NAT router
- [ ] and receives the existing blocks at start-up, storing the transparency logs in its local DB
- [ ] an inspect-log for an existing Pyrsia artifact will succeed on the regular node
- [ ] a Docker pull of an existing Pyrsia artifact will succeed
- [ ] a Docker pull or an explicit build trigger, will send a build request to the central node
- [ ] the central node listens for build requests and performs build using the build pipeline prototype
- [ ] the regular node receives blocks (transparency logs) for newly added builds
- [ ] the regular node can now inspect-log of the newly built artifact
- [ ] the regular node can now pull the newly built docker image
